File: PlaneBuilder/impl/Cockpit.js
import { Part } from "./Part.js";
import { Stats, WARNING_COLOR } from "./Stats.js";
import { BoolArr } from "./Serialize.js";
import { lu } from "./Localization.js";
export class Cockpit extends Part {
    constructor(tl, ul, sl, gl) {
        super();
        this.stats = new Stats();
        this.types = tl;
        this.upgrades = ul;
        this.safety = sl;
        this.gunsights = gl;
        this.selected_type = 0;
        this.selected_upgrades = [...Array(this.upgrades.length).fill(false)];
        this.selected_safety = [...Array(this.safety.length).fill(false)];
        this.selected_gunsights = [...Array(this.gunsights.length).fill(false)];
        this.total_stress = [0, 0];
        this.total_escape = 0;
        this.total_visibility = 0;
        this.seat_index = 0;
        this.bombsight = 0;
        this.is_armed = false;
    }
    toJSON() {
        return {
            type: this.selected_type,
            upgrades: this.selected_upgrades,
            safety: this.selected_safety,
            sights: this.selected_gunsights,
            bombsight: this.bombsight,
        };
    }
    fromJSON(js, json_version) {
        this.selected_type = js["type"];
        this.selected_upgrades = BoolArr(js["upgrades"], this.selected_upgrades.length);
        this.selected_safety = BoolArr(js["safety"], this.selected_safety.length);
        this.selected_gunsights = BoolArr(js["sights"], this.selected_gunsights.length);
        if (this.IsPrimary())
            this.selected_upgrades[0] = false;
        if (json_version > 10.35)
            this.bombsight = js["bombsight"];
    }
    serialize(s) {
        s.PushNum(this.selected_type);
        s.PushBoolArr(this.selected_upgrades);
        s.PushBoolArr(this.selected_safety);
        s.PushBoolArr(this.selected_gunsights);
        s.PushNum(this.bombsight);
    }
    deserialize(d) {
        this.selected_type = d.GetNum();
        this.selected_upgrades = d.GetBoolArr(this.selected_upgrades.length);
        this.selected_safety = d.GetBoolArr(this.selected_safety.length);
        this.selected_gunsights = d.GetBoolArr(this.selected_gunsights.length);
        if (this.IsPrimary())
            this.selected_upgrades[0] = false;
        if (d.version > 10.35)
            this.bombsight = d.GetNum();
    }
    GetTypeList() {
        return this.types;
    }
    GetUpgradeList() {
        return this.upgrades;
    }
    GetSafetyList() {
        return this.safety;
    }
    GetGunsightList() {
        return this.gunsights;
    }
    GetType() {
        return this.selected_type;
    }
    SetType(index) {
        if (index >= this.types.length)
            throw "Selected type is not in range of actual types.";
        this.selected_type = index;
        this.CalculateStats();
    }
    GetSelectedUpgrades() {
        return this.selected_upgrades;
    }
    SetUpgrade(index, state) {
        if (index >= this.upgrades.length)
            throw "Selected type is not in range of actual upgrades.";
        this.selected_upgrades[index] = state;
        this.CalculateStats();
    }
    GetSelectedSafety() {
        return this.selected_safety;
    }
    CanSafety() {
        const lst = Array(this.safety.length).fill(true);
        lst[5] = !this.types[this.selected_type].exposed;
        return lst;
    }
    SetSafety(index, state) {
        if (index >= this.safety.length)
            throw "Selected type is not in range of actual safties.";
        this.selected_safety[index] = state;
        this.CalculateStats();
    }
    GetSelectedGunsights() {
        return this.selected_gunsights;
    }
    SetGunsight(index, state) {
        if (index >= this.safety.length)
            throw "Selected type is not in range of actual gunsights.";
        this.selected_gunsights[index] = state;
        this.CalculateStats();
    }
    GetVisibility() {
        if (this.types[this.selected_type].stats.visibility < -10)
            return -1 / 0;
        return this.total_visibility;
    }
    GetFlightStress() {
        return this.total_stress;
    }
    GetEscape() {
        return this.total_escape;
    }
    GetCrash() {
        return this.total_crash;
    }
    GetAttack() {
        var mx = 0;
        for (let i = 0; i < this.gunsights.length; i++) {
            if (this.selected_gunsights[i]) {
                mx = Math.max(mx, this.gunsights[i].attack);
            }
        }
        return mx;
    }
    SetSeatIndex(idx) {
        this.seat_index = idx;
    }
    IsPrimary() {
        return this.seat_index == 0;
    }
    CanUpgrades() {
        const can = [...Array(this.upgrades.length).fill(true)];
        if (this.IsPrimary()) {
            can[0] = false;
        }
        return can;
    }
    IsOpen() {
        return this.types[this.selected_type].name == "Open"
            || this.types[this.selected_type].name == "Windscreen";
    }
    GetBombsightQuality() {
        return this.bombsight;
    }
    SetBombsightQuality(num) {
        if (num != num)
            num = 0;
        if (num == this.bombsight - 1)
            num = this.bombsight - 3;
        if (num == this.bombsight + 1)
            num = this.bombsight + 3;
        if (num < 2)
            num = 0;
        if (num > 0)
            num = num - (num % 3) + 1;
        this.bombsight = num;
        this.CalculateStats();
    }
    SetHasRotary(has) {
        this.has_rotary = has;
    }
    IsElectrics() {
        return this.stats.charge != 0;
    }
    IsCopilot() {
        return this.selected_upgrades[0];
    }
    SetArmed(is) {
        this.is_armed = is;
    }
    GetName() {
        if (this.IsPrimary()) {
            return "Crew Pilot";
        }
        if (this.bombsight > 0) {
            return "Crew Bombadier";
        }
        if (this.IsCopilot()) {
            return "Crew Co-Pilot";
        }
        if (this.is_armed) {
            return "Crew Gunner";
        }
        return "Crew Aircrew";
    }
    PartStats() {
        let stats = new Stats();
        stats.reqsections = 1;
        stats = stats.Add(this.types[this.selected_type].stats);
        for (let i = 0; i < this.selected_upgrades.length; i++) {
            if (this.selected_upgrades[i])
                stats = stats.Add(this.upgrades[i].stats);
        }
        const can = this.CanSafety();
        for (let i = 0; i < this.selected_safety.length; i++) {
            if (!can[i])
                this.selected_safety[i] = false;
            if (this.selected_safety[i])
                stats = stats.Add(this.safety[i].stats);
        }
        for (let i = 0; i < this.selected_gunsights.length; i++) {
            if (this.selected_gunsights[i])
                stats = stats.Add(this.gunsights[i].stats);
        }
        if (this.bombsight > 0) {
            stats.cost += Math.floor(1.0e-6 + 2 + (this.bombsight - 4) / 3);
            stats.warnings.push({
                source: lu("Bombsight"),
                warning: lu("Bombsight Warning 1") + this.bombsight.toString() + lu("Bombsight Warning 2"),
                color: WARNING_COLOR.WHITE,
            });
            if (this.IsCopilot()) {
                stats.warnings.push({
                    source: lu("Bombadier Controls"),
                    warning: lu("Bombadier Controls Warning"),
                    color: WARNING_COLOR.WHITE,
                });
            }
        }
        this.stats = stats.Clone();
        //Special stuff for co-pilot controls
        if (this.selected_upgrades[0]) {
            stats.flightstress = this.upgrades[0].stats.flightstress;
            this.stats.flightstress -= stats.flightstress;
        }
        else {
            stats.flightstress = 0;
        }
        return stats;
    }
    CrewUpdate(escape, controlstress, rumblestress, copilots, visibility, crash) {
        this.total_escape = this.stats.escape + escape;
        let ncp_stress = this.stats.flightstress;
        let cp_stress = this.stats.flightstress;
        if (this.IsPrimary() || this.selected_upgrades[0]) {
            ncp_stress += controlstress;
            cp_stress += controlstress - copilots * 2;
        }
        ncp_stress = Math.max(0, ncp_stress);
        cp_stress = Math.max(0, cp_stress);
        ncp_stress += rumblestress;
        cp_stress += rumblestress;
        if (this.IsOpen() && this.has_rotary) { //Is open and has rotary
            ncp_stress += 1;
            cp_stress += 1;
        }
        ncp_stress = Math.max(0, ncp_stress);
        cp_stress = Math.max(0, cp_stress);
        this.total_stress = [ncp_stress, cp_stress];
        this.total_visibility = this.stats.visibility + visibility;
        this.total_crash = this.stats.crashsafety + crash;
    }
    SetCalculateStats(callback) {
        this.CalculateStats = callback;
    }
    GetElectrics() {
        const battery_storage = 0;
        var equipment = [];
        for (let i = 0; i < this.upgrades.length; i++) {
            if (this.selected_upgrades[i]) {
                const item = this.upgrades[i];
                equipment = this.FormatEquipment(equipment, item.name, item.stats.charge);
            }
        }
        for (let i = 0; i < this.safety.length; i++) {
            if (this.selected_safety[i]) {
                const item = this.safety[i];
                equipment = this.FormatEquipment(equipment, item.name, item.stats.charge);
            }
        }
        for (let i = 0; i < this.gunsights.length; i++) {
            if (this.selected_gunsights[i]) {
                const item = this.gunsights[i];
                equipment = this.FormatEquipment(equipment, item.name, item.stats.charge);
            }
        }
        return { storage: battery_storage, equipment: equipment };
    }
}

File: flyingcircuswasm/src/cockpit.rs
use std::{iter::zip, rc::Rc};

use serde_json::Map;

use ui_core::*;
use ui_macro::*;

use crate::{
    cockpits::{CockpitType, GunsightType, SafetyType, UpgradeType},
    json::{jsboolarr, jsnum},
    lu,
    part::{ElectricsMessage, Equipment, Part},
    serialization::{JSSerializable, Serializable},
    stats::{rtz, Stats, Warning, WarningLevel},
};

#[derive(Clone, UIBindings)]
pub struct Cockpit {
    types: Rc<Vec<CockpitType>>,
    upgrades: Rc<Vec<UpgradeType>>,
    safety: Rc<Vec<SafetyType>>,
    gunsights: Rc<Vec<GunsightType>>,

    #[ui(
        select,
        source = "types",
        enabled_fn,
        enabled_opt_fn = "is_type_enabled"
    )]
    selected_type: usize,
    #[ui(check_list, source = "upgrades", enabled_fn = "is_upgrade_enabled")]
    selected_upgrades: Vec<bool>,
    #[ui(check_list, source = "safety", enabled_fn = "is_safety_enabled")]
    selected_safety: Vec<bool>,
    #[ui(check_list, source = "gunsights", enabled_fn)]
    selected_gunsights: Vec<bool>,

    this_stats: Stats,
    total_stress: (i16, i16),
    total_escape: i16,
    total_visibility: i16,
    seat_index: usize,
    #[ui(
        number,
        name = "bombsight",
        enabled_fn,
        set_fn = "set_bombsight_quality"
    )]
    bombsight: i16,
    has_rotary: bool,
    is_armed: bool,
}

impl Cockpit {
    pub fn new(
        types: &Rc<Vec<CockpitType>>,
        upgrades: &Rc<Vec<UpgradeType>>,
        safety: &Rc<Vec<SafetyType>>,
        gunsights: &Rc<Vec<GunsightType>>,
    ) -> Cockpit {
        Cockpit {
            types: types.clone(),
            upgrades: upgrades.clone(),
            safety: safety.clone(),
            gunsights: gunsights.clone(),
            selected_type: 0,
            selected_upgrades: vec![false; upgrades.len()],
            selected_safety: vec![false; safety.len()],
            selected_gunsights: vec![false; gunsights.len()],
            this_stats: Stats::new(),
            total_stress: (0, 0),
            total_escape: 0,
            total_visibility: 0,
            seat_index: 0,
            bombsight: 0,
            has_rotary: false,
            is_armed: false,
        }
    }

    fn is_primary(&self) -> bool {
        self.seat_index == 0
    }

    pub fn set_seat_index(&mut self, idx: usize) {
        self.seat_index = idx;
    }

    fn is_copilot(&self) -> bool {
        self.selected_upgrades[0]
    }

    // Enabled functions for UI bindings
    fn is_selected_type_enabled(&self) -> bool {
        true
    }

    fn is_type_enabled(&self) -> Vec<bool> {
        vec![true; self.types.len()]
    }

    fn is_upgrade_enabled(&self) -> Vec<bool> {
        // From TypeScript CanUpgrades(): can[0] = false if IsPrimary()
        let mut can = vec![true; self.upgrades.len()];
        if self.is_primary() {
            can[0] = false;
        }
        can
    }

    fn is_safety_enabled(&self) -> Vec<bool> {
        // From TypeScript CanSafety(): lst[5] = !this.types[this.selected_type].exposed
        let mut lst = vec![true; self.safety.len()];
        if self.safety.len() > 5 {
            lst[5] = !self.types[self.selected_type].exposed;
        }
        lst
    }

    fn is_selected_gunsights_enabled(&self) -> Vec<bool> {
        vec![true; self.gunsights.len()]
    }

    fn is_bombsight_enabled(&self) -> bool {
        true
    }

    /// Setter for bombsight quality with validation
    /// This implements the same logic as TypeScript's SetBombsightQuality:
    /// - Clicking up/down by 1 jumps by 3 instead
    /// - Values below 2 become 0
    /// - Values are normalized to multiples of 3 plus 1 (1, 4, 7, 10, etc.)
    fn set_bombsight_quality(&mut self, mut num: i16) {
        // Check if incrementing or decrementing by 1, jump by 3 instead
        if num == self.bombsight - 1 {
            num = self.bombsight - 3;
        }
        if num == self.bombsight + 1 {
            num = self.bombsight + 3;
        }
        // If less than 2, set to 0
        if num < 2 {
            num = 0;
        }
        // If greater than 0, normalize to multiples of 3 plus 1
        if num > 0 {
            num = num - (num % 3) + 1;
        }
        self.bombsight = num;
    }

    //TODO: On set, check CanSafety
}

impl Part for Cockpit {
    fn part_stats(&mut self) -> Stats {
        let mut s = Stats::new();
        s.reqsections += 1.0;
        s = s.add(&self.types[self.selected_type].stats);

        let _ = zip(self.upgrades.iter(), &self.selected_upgrades).map(|(val, sel)| {
            if *sel {
                s = s.add(&val.stats);
            }
        });
        let _ = zip(self.safety.iter(), &self.selected_safety).map(|(val, sel)| {
            if *sel {
                s = s.add(&val.stats);
            }
        });
        let _ = zip(self.gunsights.iter(), &self.selected_gunsights).map(|(val, sel)| {
            if *sel {
                s = s.add(&val.stats);
            }
        });

        if self.bombsight > 0 {
            s.cost += rtz(1.0e-6 + 2.0 + (self.bombsight - 4) as f64 / 3.0);
            s.warnings.push(Warning {
                name: lu!("Bombsight"),
                warning: lu!("Bombsight Warning", self.bombsight),
                level: WarningLevel::White,
            });

            if self.is_copilot() {
                s.warnings.push(Warning {
                    name: lu!("Bombadier Controls"),
                    warning: lu!("Bombadier Controls Warning"),
                    level: WarningLevel::White,
                })
            }
        }

        self.this_stats = s.clone();

        // Special stuff for co-pilot controls
        if self.is_copilot() {
            s.flightstress = self.upgrades[0].stats.flightstress;
            self.this_stats.flightstress -= s.flightstress;
        } else {
            s.flightstress = 0.0;
        }

        s
    }

    fn get_electrics(&self) -> ElectricsMessage {
        let mut equip: Vec<Equipment> = Vec::new();
        let _ = zip(self.upgrades.iter(), &self.selected_upgrades).map(|(val, sel)| {
            if *sel {
                equip.extend(format_equipment(&val.name, val.stats.charge));
            }
        });
        let _ = zip(self.safety.iter(), &self.selected_safety).map(|(val, sel)| {
            if *sel {
                equip.extend(format_equipment(&val.name, val.stats.charge));
            }
        });
        let _ = zip(self.gunsights.iter(), &self.selected_gunsights).map(|(val, sel)| {
            if *sel {
                equip.extend(format_equipment(&val.name, val.stats.charge));
            }
        });

        ElectricsMessage {
            storage: 0,
            equipment: equip,
        }
    }
}

fn format_equipment(name: &String, charge: f64) -> Vec<Equipment> {
    let mut equip: Vec<Equipment> = Vec::new();
    if charge.abs() > 0.5 {
        equip.push(Equipment {
            source: lu!(name),
            charge: charge.to_string(),
        });
    } else if charge > 0.0 && charge < 1.0e-6 {
        equip.push(Equipment {
            source: lu!(name),
            charge: "-".to_string(),
        });
    }
    equip
}

impl Serializable for Cockpit {
    fn deserialize(
        &mut self,
        d: &mut crate::serialization::Deserializer,
    ) -> Result<(), crate::serialization::Error> {
        self.selected_type = d.get_num()? as usize;
        self.selected_upgrades = d.get_bool_arr(self.upgrades.len())?;
        self.selected_safety = d.get_bool_arr(self.safety.len())?;
        self.selected_gunsights = d.get_bool_arr(self.gunsights.len())?;
        if self.is_primary() {
            self.selected_upgrades[0] = false;
        }
        if d.version > 10.35 {
            self.bombsight = d.get_num()?;
        }
        Ok(())
    }

    fn serialize(
        &self,
        s: &mut crate::serialization::Serializer,
    ) -> Result<(), crate::serialization::Error> {
        s.push_num(self.selected_type as i16)?;
        s.push_bool_arr(&self.selected_upgrades)?;
        s.push_bool_arr(&self.selected_safety)?;
        s.push_bool_arr(&self.selected_gunsights)?;
        s.push_num(self.bombsight)?;
        Ok(())
    }
}

impl JSSerializable for Cockpit {
    fn from_json(&mut self, js: &serde_json::Value, json_version: f64) {
        self.selected_type = jsnum(js, "type") as usize;
        self.selected_upgrades = jsboolarr(js, "upgrades");
        self.selected_safety = jsboolarr(js, "safety");
        self.selected_gunsights = jsboolarr(js, "sights");
        if self.is_primary() {
            self.selected_upgrades[0] = false;
        }
        if json_version > 10.35 {
            self.bombsight = jsnum(js, "bombsight") as i16;
        }
    }
    fn to_json(&self) -> serde_json::Value {
        let mut map = Map::new();
        map.insert("type".to_string(), self.selected_type.into());
        map.insert(
            "upgrades".to_string(),
            serde_json::Value::Array(
                self.selected_upgrades
                    .iter()
                    .map(|b| serde_json::Value::Bool(*b))
                    .collect(),
            ),
        );
        map.insert(
            "safety".to_string(),
            serde_json::Value::Array(
                self.selected_safety
                    .iter()
                    .map(|b| serde_json::Value::Bool(*b))
                    .collect(),
            ),
        );
        map.insert(
            "sights".to_string(),
            serde_json::Value::Array(
                self.selected_gunsights
                    .iter()
                    .map(|b| serde_json::Value::Bool(*b))
                    .collect(),
            ),
        );
        map.insert("bombsight".to_string(), self.bombsight.into());

        serde_json::Value::Object(map)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_set_bombsight_quality() {
        let types = Rc::new(vec![]);
        let upgrades = Rc::new(vec![]);
        let safety = Rc::new(vec![]);
        let gunsights = Rc::new(vec![]);
        let mut cockpit = Cockpit::new(&types, &upgrades, &safety, &gunsights);

        // Test setting to 0
        cockpit.bombsight = 0;
        cockpit.set_bombsight_quality(0);
        assert_eq!(cockpit.bombsight, 0, "0 should stay 0");

        // Test values below 2 become 0 when not adjacent
        cockpit.bombsight = 10; // Set to a non-adjacent value first
        cockpit.set_bombsight_quality(1);
        assert_eq!(cockpit.bombsight, 0, "1 should become 0 when not adjacent");

        // Test normalization to multiples of 3 plus 1 (when not adjacent)
        cockpit.bombsight = 0; // Reset to non-adjacent
        cockpit.set_bombsight_quality(4);
        assert_eq!(cockpit.bombsight, 4, "4 should stay 4 (4 = 3*1 + 1)");

        // Test normalization: num - (num % 3) + 1
        // Valid values are 1, 4, 7, 10, 13, ... (multiples of 3 plus 1)
        cockpit.bombsight = 10; // Set to non-adjacent first
        cockpit.set_bombsight_quality(5);
        assert_eq!(cockpit.bombsight, 4, "5 normalizes to 4 (5 - 2 + 1)");

        cockpit.bombsight = 10;
        cockpit.set_bombsight_quality(6);
        assert_eq!(cockpit.bombsight, 7, "6 normalizes to 7 (6 - 0 + 1)");

        cockpit.bombsight = 10;
        cockpit.set_bombsight_quality(7);
        assert_eq!(cockpit.bombsight, 7, "7 stays 7 (7 = 3*2 + 1)");

        cockpit.bombsight = 10;
        cockpit.set_bombsight_quality(8);
        assert_eq!(cockpit.bombsight, 7, "8 normalizes to 7 (8 - 1 + 1)");

        cockpit.bombsight = 10;
        cockpit.set_bombsight_quality(9);
        assert_eq!(cockpit.bombsight, 7, "9 normalizes to 7 (9 - 2 + 1)");

        cockpit.bombsight = 10;
        cockpit.set_bombsight_quality(10);
        assert_eq!(cockpit.bombsight, 10, "10 stays 10 (10 - 0 + 1)");

        // Test incrementing by 1 (arrow key) jumps by 3
        cockpit.bombsight = 4;
        cockpit.set_bombsight_quality(5); // Detected as +1
        assert_eq!(cockpit.bombsight, 7, "4+1 should jump to 7");

        // Test decrementing by 1 (arrow key) jumps by 3
        cockpit.bombsight = 7;
        cockpit.set_bombsight_quality(6); // Detected as -1
        assert_eq!(cockpit.bombsight, 4, "7-1 should jump to 4");

        // Test decrementing from 4 to 3 jumps to 1, which becomes 0
        cockpit.bombsight = 4;
        cockpit.set_bombsight_quality(3); // Detected as -1
        assert_eq!(cockpit.bombsight, 0, "4-1 jumps to 1, which becomes 0");

        // Test incrementing from 0 by 1 jumps to 3, normalizes to 4
        cockpit.bombsight = 0;
        cockpit.set_bombsight_quality(1); // Detected as +1
        assert_eq!(cockpit.bombsight, 4, "0+1 jumps to 3, normalizes to 4");
    }
}
